themes/V3/5ePHB/snippets/tableOfContents.gen.js

```diff
@@ -28,31 +28,33 @@ const getTOC = (pages)=>{
 	};
 
 	const res = [];
-	_.each(pages, (page, pageNum)=>{
-		if(!page.includes("{{frontCover}}") && !page.includes("{{insideCover}}") && !page.includes("{{partCover}}") && !page.includes("{{backCover}}")) {
-			const lines = page.split('\n');
-			_.each(lines, (line)=>{
-				if(_.startsWith(line, '# ')){
-					const title = line.replace('# ', '');
-					add1(title, pageNum);
+	const iframe = document.getElementById('BrewRenderer');
+	const iframeDocument = iframe.contentDocument || iframe.contentWindow.document;
+	const headings = iframeDocument.querySelectorAll('h1, h2, h3');
+
+	_.each(headings, (heading)=>{
+		const onPage = parseInt(heading.closest('.page,.phb').id?.replace(/^p/, ''));
+		const ToCExclude = getComputedStyle(heading).getPropertyValue('--TOC');
+		if(ToCExclude != '"exclude"') {
+			if(!isNaN(onPage)) {
```

Comment:
Page numbers appear to be off-by-1 (items on page 2 are labeled page 3, etc.)

Reply:
Are you looking at physical page numbers, or logical page numbers?

Reply:
The first page of the document is page 1. Clicking the links in the TOC takes you to the page after.

For example, if I have

`# Chapter 1` on page 2, the TOC will say it is page 3, even though I dont have a page 3.

Reply:
I'm seeing something different, but also wrong.

My document is

1. front cover
2. table of contents
3. some page 3
4. another page 4

but .. the TOC looks like this:

```
- ### [{{ GATE THUNDER}}{{ 2}}](#p2)
  - #### [{{ Shadowy Depths}}{{ 2}}](#p2)
- ### [{{ Table Of Contents is on Physical Page 2}}{{ 3}}](#p3)
- ### [{{ Physical Page 3 Heading 1}}{{ 4}}](#p4)
  - #### [{{ Heading 2}}{{ 4}}](#p4)
  - #### [{{ Heading 2}}{{ 4}}](#p4)
    - [{{ a very long heading that might wrap very poorly}}{{ 4}}](#p4)
- ### [{{ Physical Page 4}}{{ 5}}](#p5)
  - #### [{{ Sparklemuffin Peacock Spider}}{{ 5}}](#p5)
- ### [{{ Physical Page 5}}{{ 6}}](#p6)
  - #### [{{ Heading 1}}{{ 6}}](#p6)
  - #### [{{ Heading 2}}{{ 6}}](#p6)
  - #### [{{ Heading 2}}{{ 6}}](#p6)
- ### [{{ Physical Page 6}}{{ 7}}](#p7)
  - #### [{{ Heading 1}}{{ 7}}](#p7)
  - #### [{{ Heading 2}}{{ 7}}](#p7)
  - #### [{{ Heading 2}}{{ 7}}](#p7)
```

Here, each _displayed_ page number does correspond to the _linked_ page number .. however the page _number_ value in both cases is off by 1.

Reply:
This is the same thing I'm seeing. The generated link and displayed number are equal, but they are both wrong by 1.

Reply:
The `onPage = parseInt(heading.closest('.page,.phb').id?.replace(/^p/, ''))` step is working correctly (tested via Store as Global Variable in console.).

That value is passed in to functions `add1`, `add2`, `add3`.

Looking at those functions reveals the problem:
```
	const add1 = (title, page)=>{
		res.push({
			title    : title,
			page     : page + 1, // BUG IS HERE
			children : []
		});
	};
```

I believe this is a hold over from when toc entries were found by searching through raw brew text that had been split into pages into a zero-indexed array. The page number is now being calculated from the `id="pNNN"` of the nearest `<div class="page">` element.